File: src/main/java/com/brasil/transparente/api/controller/FinderController.java
package com.brasil.transparente.api.controller;

import com.brasil.transparente.api.dto.DisplayableElementDTO;
import com.brasil.transparente.api.entity.DespesaSimplificada;
import com.brasil.transparente.api.service.FinderService;
import io.swagger.v3.oas.annotations.OpenAPIDefinition;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.info.Info;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.AllArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.util.List;

import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;

@CrossOrigin(origins = "*", methods = {RequestMethod.GET})
@OpenAPIDefinition(
        info = @Info(
                title = "Brasil Transparente REST API",
                version = "1.0.0-2024",
                description = "Canonical API for exploration of effectively paid Brazilian public spending at federal and state levels. "
                        + "This API exposes two strictly separated data perspectives: Summary endpoints for high-level aggregated analysis, "
                        + "and Exploration endpoints for hierarchical drill-down. "
                        + "All identifiers MUST be obtained dynamically from API responses and MUST NOT be inferred, guessed, or hard-coded. "
                        + "The OpenAPI specification is the single source of truth for structure, hierarchy, and behavior."
        )
)
@RestController
@AllArgsConstructor
public class FinderController {

    private final FinderService finderService;

    @Tag(
            name = "Exploration",
            description = "Hierarchical drill-down endpoints that MUST be used strictly in the documented order. IDs obtained at one level are valid ONLY for the immediately next level."
    )
    @Operation(
            summary = "ENTRY POINT – List Powers of a Federative Unit",
            description = "MANDATORY ENTRY POINT for hierarchical exploration. Returns Executive, Legislative, Judicial, and other powers. "
                    + "For Brazilian Federal Government analysis, unidadeFederativaId MUST be 1. All hierarchical exploration MUST start from this endpoint."
    )
    @GetMapping(value = "/unidade-federativa/{unidadeFederativaId}/poderes", produces = APPLICATION_JSON_VALUE)
    public List<DisplayableElementDTO> getPoderesByUnidadeFederativa(
            @Parameter(description = "Federative Unit identifier. Use 1 for the Brazilian Federal Government. This value is the mandatory starting point for exploration.", required = true)
            @PathVariable Long unidadeFederativaId
    ) {
        return finderService.getPoderesByUnidadeFederativa(unidadeFederativaId);
    }

    @Tag(name = "Exploration")
    @Operation(
            summary = "List Ministries by Power",
            description = "Returns ministries or courts associated with a specific Power. This endpoint MUST only be called after obtaining a valid Power ID from /unidade-federativa/{id}/poderes. IDs from any other hierarchy level are INVALID."
    )
    @GetMapping(value = "/poder/{idPoder}/ministerios", produces = APPLICATION_JSON_VALUE)
    public List<DisplayableElementDTO> getMinisterioByPoderId(
            @Parameter(description = "Power identifier obtained EXCLUSIVELY from the powers endpoint.", required = true)
            @PathVariable Long idPoder
    ) {
        return finderService.getMinisterioByPoderId(idPoder);
    }

    @Tag(name = "Exploration")
    @Operation(
            summary = "List Agencies by Ministry",
            description = "Returns agencies associated with a specific Ministry. This endpoint MUST only be called after obtaining a Ministry ID from /poder/{idPoder}/ministerios. Skipping hierarchy levels or reusing IDs from other contexts is NOT allowed."
    )
    @GetMapping(value = "/ministerio/{idMinisterio}/orgaos", produces = APPLICATION_JSON_VALUE)
    public List<DisplayableElementDTO> getOrgaoByMinisterioId(
            @Parameter(description = "Ministry identifier obtained EXCLUSIVELY from the ministries endpoint.", required = true)
            @PathVariable Long idMinisterio
    ) {
        return finderService.getOrgaoByMinisterioId(idMinisterio);
    }

    @Tag(name = "Exploration")
    @Operation(
            summary = "List Management Units by Agency",
            description = "Returns Management Units responsible for budget execution within a specific Agency. This endpoint MUST only be called after obtaining an Agency ID from /ministerio/{idMinisterio}/orgaos."
    )
    @GetMapping(value = "/orgao/{idOrgao}/unidades-gestoras", produces = APPLICATION_JSON_VALUE)
    public List<DisplayableElementDTO> getUnidadeGestoraByOrgaoId(
            @Parameter(description = "Agency identifier obtained EXCLUSIVELY from the agencies endpoint.", required = true)
            @PathVariable Long idOrgao
    ) {
        return finderService.getUnidadeGestoraByOrgaoId(idOrgao);
    }

    @Tag(name = "Exploration")
    @Operation(
            summary = "FINAL LEVEL – List Expense Elements by Management Unit",
            description = "Final level of the hierarchical exploration. Returns detailed paid expense elements for a Management Unit. "
                    + "This endpoint MUST only be called after obtaining a Management Unit ID from /orgao/{idOrgao}/unidades-gestoras. "
                    + "No further drill-down is possible beyond this level."
    )
    @GetMapping(value = "/unidade-gestora/{idUnidadeGestora}/elemento-despesa", produces = APPLICATION_JSON_VALUE)
    public List<DisplayableElementDTO> getElementoDespesaByUnidadeGestoraId(
            @Parameter(description = "Management Unit identifier obtained EXCLUSIVELY from the management units endpoint.", required = true)
            @PathVariable Long idUnidadeGestora
    ) {
        return finderService.getElementoDespesaByUnidadeGestoraId(idUnidadeGestora);
    }

    @Tag(
            name = "Summary",
            description = "High-level aggregated views for macro analysis. Summary data MUST NOT be combined or re-aggregated with Exploration endpoints."
    )
    @Operation(
            summary = "Simplified Spending Report by Federative Unit",
            description = "High-level reclassified aggregation of public spending categories based on project methodology. "
                    + "This endpoint is intended ONLY for macro analysis and narrative explanations. "
                    + "Results from this endpoint MUST NOT be merged or aggregated with Exploration endpoints."
    )
    @GetMapping(value = "/despesa-simplificada/{unidadeFederativaId}", produces = APPLICATION_JSON_VALUE)
    public List<DespesaSimplificada> getSimplifiedReport(
            @Parameter(description = "Federative Unit identifier. Use 1 for the Brazilian Federal Government.", required = true)
            @PathVariable Long unidadeFederativaId
    ) {
        return finderService.getSimplifiedReport(unidadeFederativaId);
    }

    @Tag(name = "Summary")
    @Operation(
            summary = "Total Paid Spending by Federative Unit",
            description = "Total amount of effectively paid public spending for a federative unit. This value represents a cash-based annual snapshot and MUST NOT be decomposed using Exploration endpoints."
    )
    @GetMapping(value = "/unidade-federativa/{unidadeFederativaId}/total-value-spent", produces = APPLICATION_JSON_VALUE)
    public Double getTotalValueSpentByUnidadeFederativaId(
            @Parameter(description = "Federative Unit identifier. Use 1 for the Brazilian Federal Government.", required = true)
            @PathVariable Long unidadeFederativaId
    ) {
        return finderService.getTotalValueSpentByUnidadeFederativaId(unidadeFederativaId);
    }
}

File: src/main/kotlin/com/brasil/transparente/api/controller/FinderController.kt
package com.brasil.transparente.api.controller

import com.brasil.transparente.api.dto.DisplayableElementDTO
import com.brasil.transparente.api.entity.DespesaSimplificada
import com.brasil.transparente.api.service.FinderService
import io.swagger.v3.oas.annotations.OpenAPIDefinition
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.info.Info
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.web.bind.annotation.*
import org.springframework.http.MediaType.APPLICATION_JSON_VALUE

@CrossOrigin(origins = ["*"], methods = [RequestMethod.GET])
@OpenAPIDefinition(
    info = Info(
        title = "Brasil Transparente REST API",
        version = "1.0.0-2024",
        description = "API para exploração de gastos públicos brasileiros efetivamente pagos em nível federal. " +
                "Esta API expõe duas perspectivas de dados estritamente separadas: endpoints de Resumo para análise agregada de alto nível e endpoints de Exploração para navegação hierárquica detalhada. " +
                "Todos os identificadores devem ser obtidos dinamicamente a partir das respostas da API e não devem ser inferidos, adivinhados ou fixados no código. "
    )
)
@RestController
class FinderController(private val finderService: FinderService) {

    @Tag(
        name = "Exploration",
        description = "Endpoints de navegação hierárquica que devem ser usados estritamente na ordem documentada. IDs obtidos em um nível são válidos somente para o nível imediatamente seguinte."
    )
    @Operation(
        summary = "Ponto de entrada – Lista os Poderes de uma Unidade Federativa",
        description = "Ponto de entrada obrigatório para exploração hierárquica. Retorna os poderes Executivo, Legislativo, Judiciário e outros. " +
                "Para análise do Governo Federal brasileiro, unidadeFederativaId deve ser 1. Toda exploração hierárquica deve começar por este endpoint."
    )
    @GetMapping(value = ["/unidade-federativa/{unidadeFederativaId}/poderes"], produces = [APPLICATION_JSON_VALUE])
    fun getPoderesByUnidadeFederativa(
        @Parameter(description = "Identificador da Unidade Federativa. Use 1 para o Governo Federal brasileiro. Este valor é o ponto de partida obrigatório para a exploração.", required = true)
        @PathVariable unidadeFederativaId: Long
    ): List<DisplayableElementDTO> = finderService.getPoderesByUnidadeFederativa(unidadeFederativaId)

    @Tag(name = "Exploration")
    @Operation(
        summary = "Lista Ministérios por Poder",
        description = "Retorna ministérios ou tribunais associados a um Poder específico. Este endpoint deve ser chamado apenas após obter um ID de Poder válido em /unidade-federativa/{id}/poderes. IDs de qualquer outro nível hierárquico são inválidos."
    )
    @GetMapping(value = ["/poder/{idPoder}/ministerios"], produces = [APPLICATION_JSON_VALUE])
    fun getMinisterioByPoderId(
        @Parameter(description = "Identificador do Poder obtido exclusivamente do endpoint de poderes.", required = true)
        @PathVariable idPoder: Long
    ): List<DisplayableElementDTO> = finderService.getMinisterioByPoderId(idPoder)

    @Tag(name = "Exploration")
    @Operation(
        summary = "Lista Órgãos por Ministério",
        description = "Retorna órgãos associados a um Ministério específico. Este endpoint deve ser chamado apenas após obter um ID de Ministério em /poder/{idPoder}/ministerios. Pular níveis da hierarquia ou reutilizar IDs de outros contextos não é permitido."
    )
    @GetMapping(value = ["/ministerio/{idMinisterio}/orgaos"], produces = [APPLICATION_JSON_VALUE])
    fun getOrgaoByMinisterioId(
        @Parameter(description = "Identificador do Ministério obtido exclusivamente do endpoint de ministérios.", required = true)
        @PathVariable idMinisterio: Long
    ): List<DisplayableElementDTO> = finderService.getOrgaoByMinisterioId(idMinisterio)

    @Tag(name = "Exploration")
    @Operation(
        summary = "Lista Unidades Gestoras por Órgão",
        description = "Retorna Unidades Gestoras responsáveis pela execução orçamentária dentro de um Órgão específico. Este endpoint deve ser chamado apenas após obter um ID de Órgão em /ministerio/{idMinisterio}/orgaos."
    )
    @GetMapping(value = ["/orgao/{idOrgao}/unidades-gestoras"], produces = [APPLICATION_JSON_VALUE])
    fun getUnidadeGestoraByOrgaoId(
        @Parameter(description = "Identificador do Órgão obtido exclusivamente do endpoint de órgãos.", required = true)
        @PathVariable idOrgao: Long
    ): List<DisplayableElementDTO> = finderService.getUnidadeGestoraByOrgaoId(idOrgao)

    @Tag(name = "Exploration")
    @Operation(
        summary = "Nível final – Lista Elementos de Despesa por Unidade Gestora",
        description = "Nível final da exploração hierárquica. Retorna elementos de despesa detalhados e efetivamente pagos por uma Unidade Gestora. " +
                "Este endpoint deve ser chamado apenas após obter um ID de Unidade Gestora em /orgao/{idOrgao}/unidades-gestoras. " +
                "Não é possível aprofundar a exploração além deste nível."
    )
    @GetMapping(value = ["/unidade-gestora/{idUnidadeGestora}/elemento-despesa"], produces = [APPLICATION_JSON_VALUE])
    fun getElementoDespesaByUnidadeGestoraId(
        @Parameter(description = "Identificador da Unidade Gestora obtido exclusivamente do endpoint de unidades gestoras.", required = true)
        @PathVariable idUnidadeGestora: Long
    ): List<DisplayableElementDTO> = finderService.getElementoDespesaByUnidadeGestoraId(idUnidadeGestora)

    @Tag(
        name = "Summary",
        description = "Visões agregadas de alto nível para análise macro. Dados de Resumo não devem ser combinados ou reagregados com endpoints de Exploração."
    )
    @Operation(
        summary = "Relatório Simplificado de Gastos por Unidade Federativa",
        description = "Agregação reclassificada de alto nível das categorias de gasto público, com base na metodologia do projeto. " +
                "Este endpoint é destinado somente a análises macro e explicações narrativas. " +
                "Resultados deste endpoint não devem ser combinados ou agregados com endpoints de Exploração."
    )
    @GetMapping(value = ["/despesa-simplificada/{unidadeFederativaId}"], produces = [APPLICATION_JSON_VALUE])
    fun getDespesaSimplificada(
        @Parameter(description = "Identificador da Unidade Federativa. Use 1 para o Governo Federal brasileiro.", required = true)
        @PathVariable unidadeFederativaId: Long
    ): List<DespesaSimplificada> = finderService.getDespesaSimplificada(unidadeFederativaId)

    @Tag(name = "Summary")
    @Operation(
        summary = "Total de Gastos Pagos por Unidade Federativa",
        description = "Valor total de gastos públicos efetivamente pagos por uma unidade federativa. Este valor representa um retrato anual em regime de caixa e não deve ser decomposto usando endpoints de Exploração."
    )
    @GetMapping(value = ["/unidade-federativa/{unidadeFederativaId}/total-value-spent"], produces = [APPLICATION_JSON_VALUE])
    fun getTotalValueSpentByUnidadeFederativaId(
        @Parameter(description = "Identificador da Unidade Federativa. Use 1 para o Governo Federal brasileiro.", required = true)
        @PathVariable unidadeFederativaId: Long
    ): Double = finderService.getTotalValueSpentByUnidadeFederativaId(unidadeFederativaId)
}